### PR TITLE
Removed IRelease Interface

### PR DESCRIPTION
**Pull request by apankratovantonp**

### CODE REVIEW

**Comment by apankratovantonp** [comment]
Waiting for https://github.com/openvinotoolkit/openvino/pull/4258

**Comment by alpkn** [comment]
@apankratovantonp 
Would you summarize the changes and the rationale behind the committed code please?

File: inference-engine/tests_deprecated/unit/engines/mkldnn/graph/layers/internal/graph_tile_test.cpp [hunk 1]
```diff
@@ -215,8 +215,8 @@ class MKLDNNGraphDynBatchTileTests: public MKLDNNGraphTileTests {
             InferenceEngine::CNNNetwork network;
             ASSERT_NO_THROW(network = core.ReadNetwork(model, InferenceEngine::Blob::CPtr()));
 
-            auto implNet = dynamic_cast<InferenceEngine::details::CNNNetworkImpl *>(&((InferenceEngine::ICNNNetwork&)network));
-            ASSERT_NE(nullptr, implNet) << "Failed to cast ICNNNetwork to CNNNetworkImpl";
+            ASSERT_EQ(nullptr, network.getFunction());
+            auto implNet = static_cast<InferenceEngine::details::CNNNetworkImpl *>(&((InferenceEngine::ICNNNetwork&)network));
```
Comment (ilya-lavrenov) [inline inference-engine/tests_deprecated/unit/engines/mkldnn/graph/layers/internal/graph_tile_test.cpp:219]: why do we need to change it?

File: inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp [hunk 2]
```diff
@@ -65,12 +65,6 @@ CNNNetwork ONNXReader::read(std::istream& model, const std::vector<IExtensionPtr
...
-INFERENCE_PLUGIN_API(StatusCode) InferenceEngine::CreateReader(IReader*& reader, ResponseDesc *resp) noexcept {
-    try {
-        reader = new ONNXReader();
-        return OK;
-    }
-    catch (std::exception &) {
-        return GENERAL_ERROR;
-    }
+INFERENCE_PLUGIN_API(void) InferenceEngine::CreateReader(std::shared_ptr<IReader>& reader) {
+    reader = std::make_shared<ONNXReader>();
```
Comment (ilya-lavrenov) [inline inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp:69]: can we return as return value?

File: inference-engine/src/preprocessing/ie_preprocess_data.hpp [hunk 3]
```diff
@@ -58,9 +58,12 @@ class IPreProcessData : public details::IRelease {
 
     //FIXME: rename to verifyAplicable
     virtual void isApplicable(const Blob::Ptr &src, const Blob::Ptr &dst) = 0;
+
+protected:
+    ~IPreProcessData() = default;
```
Comment (ilya-lavrenov) [inline inference-engine/src/preprocessing/ie_preprocess_data.hpp:63]: why is it protected and not virtual?

File: inference-engine/src/plugin_api/cpp_interfaces/base/ie_executable_network_base.hpp [hunk 4]
```diff
@@ -87,8 +87,10 @@ class ExecutableNetworkBase : public IExecutableNetwork {
     }
     IE_SUPPRESS_DEPRECATED_END
 
-    void Release() noexcept override {
-        delete this;
+    /// @private Need for unit tests only - TODO: unit tests should test using public API, non having details
+    template<typename T>
+    const std::shared_ptr<T> getImpl() const {
+        return _impl;
     }
```
Comment (ilya-lavrenov) [inline inference-engine/src/plugin_api/cpp_interfaces/base/ie_executable_network_base.hpp:94]: do we need this?

File: inference-engine/src/inference_engine/os/lin/lin_shared_object_loader.cpp [hunk 5]
```diff
@@ -7,6 +7,7 @@
 #include "details/ie_exception.hpp"
 #include "details/ie_so_loader.h"
 #include "file_utils.h"
+#include <iostream>
```
Comment (ilya-lavrenov) [inline inference-engine/src/inference_engine/os/lin/lin_shared_object_loader.cpp:10]: debug?

File: inference-engine/include/ie_iextension.h [hunk 6]
```diff
@@ -198,9 +207,7 @@ using IExtensionPtr = std::shared_ptr<IExtension>;
  * @brief Creates the default instance of the extension
  *
  * @param ext Extension interface
- * @param resp Response description
- * @return Status code
  */
-INFERENCE_EXTENSION_API(StatusCode) CreateExtension(IExtension*& ext, ResponseDesc* resp) noexcept;
+INFERENCE_EXTENSION_API(void) CreateExtension(IExtensionPtr& ext);
```
Comment (ilya-lavrenov) [inline inference-engine/include/ie_iextension.h:211]: how will it work for existing extensions? I suppose it will break compatibility..

File: docs/template_extension/extension.cpp [hunk 7]
```diff
@@ -110,17 +110,7 @@ InferenceEngine::ILayerImpl::Ptr Extension::getImplementation(const std::shared_
...
-    try {
-        ext = new Extension();
-        return OK;
-    } catch (std::exception &ex) {
-        if (resp) {
-            std::string err = ((std::string) "Couldn't create extension: ") + ex.what();
-            err.copy(resp->msg, 255);
-        }
-        return InferenceEngine::GENERAL_ERROR;
-    }
```
Comment (ilya-lavrenov) [inline docs/template_extension/extension.cpp:124]: we need to support an old way of extensions creation for some time..

File: inference-engine/tests_deprecated/unit/engines/mkldnn/graph/layers/internal/graph_tile_test.cpp (hunk 1 above)
Comment (apankratovantonp) [reply inference-engine/tests_deprecated/unit/engines/mkldnn/graph/layers/internal/graph_tile_test.cpp:219]: `network.getFunction()` is an official IE way to check implementation. We can do the same with `dynamic_cast`. Also I have removed virtual from destructor of base class (as we used `std::shared_ptr`) so there is no RTTI with clang.

File: inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp (hunk 2 above)
Comment (apankratovantonp) [reply inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp:69]: No. We should use C API here (without mangling). It Dose not support such type as `std::shared_ptr` as return type.

File: inference-engine/src/preprocessing/ie_preprocess_data.hpp (hunk 3 above)
Comment (apankratovantonp) [reply inference-engine/src/preprocessing/ie_preprocess_data.hpp:63]: As this and all others  entities in IE are loaded using `std::shared_ptr` (as they are created inside dynamic library and we should provide deleter from library too) we do not use deletion by base reference.  We protect from base destructor call from external code. Virtual just not needed here.

File: inference-engine/src/plugin_api/cpp_interfaces/base/ie_executable_network_base.hpp (hunk 4 above)
Comment (apankratovantonp) [reply inference-engine/src/plugin_api/cpp_interfaces/base/ie_executable_network_base.hpp:94]: As comment say it is used in tests

File: inference-engine/src/inference_engine/os/lin/lin_shared_object_loader.cpp (hunk 5 above)
Comment (apankratovantonp) [reply inference-engine/src/inference_engine/os/lin/lin_shared_object_loader.cpp:10]: will be removed

File: inference-engine/include/ie_iextension.h (hunk 6 above)
Comment (apankratovantonp) [reply inference-engine/include/ie_iextension.h:211]: Seems you right

**Comment by ilya-lavrenov** [comment]
BTW, do we need `no_copy`? seems it can be removed as well.

File: inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp (hunk 2 above)
Comment (ilya-lavrenov) [reply inference-engine/src/readers/onnx_reader/ie_onnx_reader.cpp:69]: I have compiled:
```
INFERENCE_EXTENSION_API(std::shared_ptr<InferenceEngine::IExtension>) InferenceEngine::CreateExtensionShared() {
    return std::make_shared<Extension>();
}
```
And I have:
```
$ readelf -Ws lib/libtemplate_extension.so | grep CreateExtension
   167: 0000000000010220   211 FUNC    GLOBAL DEFAULT   11 CreateExtensionShared
   198: 00000000000101b0   112 FUNC    GLOBAL DEFAULT   11 CreateExtension
```

PS: Not sure about Windows..

File: docs/template_extension/extension.cpp [hunk 8]
```diff
@@ -108,6 +108,11 @@ InferenceEngine::ILayerImpl::Ptr Extension::getImplementation(const std::shared_
 }
 //! [extension:getImplementation]
 
+// Exported function
+INFERENCE_EXTENSION_API(void) InferenceEngine::CreateExtensionShared(std::shared_ptr<InferenceEngine::IExtension>& ext) {
+    ext = std::make_shared<Extension>();
+}
```
Comment (ilya-lavrenov) [inline docs/template_extension/extension.cpp:114]: When I commented this function implementation, I have:
```
&strContains(ex.what(), "name: custom_opset. Opset") evaluates to false, where
ex.what() evaluates to bad_weak_ptr
"name: custom_opset. Opset" evaluates to name: custom_opset. Opset
[WARN] 2021-02-20T13:32:32z frontend/onnx_import/src/ops_bridge.cpp 201 Overwriting existing operator: custom_domain.Template:1
[WARN] 2021-02-20T13:32:32z frontend/onnx_import/src/ops_bridge.cpp 201 Overwriting existing operator: custom_domain.FFT:1
/home/sandye51/Documents/Programming/git_repo/dldt/inference-engine/tests/functional/inference_engine/core_threading_tests.cpp:54: Failure
&strContains(ex.what(), "name: custom_opset. Opset") evaluates to false, where
ex.what() evaluates to bad_weak_ptr
"name: custom_opset. Opset" evaluates to name: custom_opset. Opset
[  FAILED  ] CoreThreadingTests.ReadNetwork (100 ms)
```

Looks like compatibility path does not work properly

File: inference-engine/include/ie_iextension.h [hunk 9]
```diff
@@ -187,6 +186,15 @@ class INFERENCE_ENGINE_API_CLASS(IExtension) : public InferenceEngine::details::
      * @param versionInfo Pointer to version info, will be set by plugin
      */
     virtual void GetVersion(const InferenceEngine::Version*& versionInfo) const noexcept = 0;
+
+    /**
+     * @brief Implements deprecated API
+     */
+    // INFERENCE_ENGINE_DEPRECATED("Will do nothing. Do not use it")
```
Comment (ilya-lavrenov) [inline inference-engine/include/ie_iextension.h:193]: can we keep `INFERENCE_ENGINE_DEPRECATED` here and for `make_so_pointer`?

File: docs/template_extension/extension.cpp (hunk 8 above)
Comment (ilya-lavrenov) [reply docs/template_extension/extension.cpp:114]: Easy reproduced with `CustomOpsSerializationTest.CustomOpUser_MO` test

**Comment by apankratovantonp** [comment]
> BTW, do we need `no_copy`? seems it can be removed as well.
It is used in some plugins. So it is better to move it deletion to separate PR

File: docs/template_extension/extension.cpp (hunk 8 above)
Comment (apankratovantonp) [reply docs/template_extension/extension.cpp:114]: What binary run this test?

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
@ilyachur @vinograd47 could you please have a look? The PR seems to contain a lot of changes, but they are straightforward. The meaningful changes are in:
1. Creation extensions
2. SOPointer loader

File: inference-engine/include/ie_icnn_network.hpp [hunk 10]
```diff
@@ -216,9 +215,7 @@ class INFERENCE_ENGINE_ICNNNETWORK_CLASS(ICNNNetwork) : public details::IRelease
         return NOT_IMPLEMENTED;
     }
 
-    /**
-     * @brief A virtual destructor.
-     */
-    virtual ~ICNNNetwork();
+protected:
+    ~ICNNNetwork() = default;
```
Comment (ilyachur) [inline inference-engine/include/ie_icnn_network.hpp:219]: Why did you remove virtual destructor?

File: inference-engine/include/ie_extension.h [hunk 11]
```diff
@@ -105,24 +100,32 @@ class INFERENCE_ENGINE_API_CLASS(Extension) : public IExtension {
...
- * @return shared_pointer A wrapper for the given type from a specific shared module
+ * @brief Creates extension using deprecated API
+ * @tparam T extension type
+ * @param name extension library name
+ * @return shared pointer to extension
  */
-template <>
-inline std::shared_ptr<IExtension> make_so_pointer(const std::string& name) {
+template<typename T = IExtension>
+// INFERENCE_ENGINE_DEPRECATED("Use std::make_shared<Extension>")
```
Comment (ilyachur) [inline inference-engine/include/ie_extension.h:111]: Why did you add deprecation macros as comments?

File: inference-engine/include/ie_extension.h [hunk 12]
```diff
@@ -105,24 +100,32 @@ class INFERENCE_ENGINE_API_CLASS(Extension) : public IExtension {
...
+ */
+template<typename T = IExtension>
+// INFERENCE_ENGINE_DEPRECATED("Use std::make_shared<Extension>")
 inline std::shared_ptr<IExtension> make_so_pointer(const std::wstring& name) {
     return std::make_shared<Extension>(name);
 }
 
 #endif
-
+// IE_SUPPRESS_DEPRECATED_END
```
Comment (ilyachur) [inline inference-engine/include/ie_extension.h:130]: I don't see the reason why we need this comment

File: inference-engine/include/ie_iexecutable_network.hpp [hunk 13]
```diff
@@ -173,6 +173,9 @@ class IExecutableNetwork : public details::IRelease {
      * @return code of the operation. InferenceEngine::OK if succeeded
      */
     virtual StatusCode GetContext(RemoteContext::Ptr& pContext, ResponseDesc* resp) const noexcept = 0;
+
+protected:
+    ~IExecutableNetwork() = default;
```
Comment (ilyachur) [inline inference-engine/include/ie_iexecutable_network.hpp:178]: Maybe do we need to have virtual destructor here?

File: inference-engine/include/ie_iextension.h [hunk 14]
```diff
@@ -187,6 +186,17 @@ class INFERENCE_ENGINE_API_CLASS(IExtension) : public InferenceEngine::details::
      * @param versionInfo Pointer to version info, will be set by plugin
      */
     virtual void GetVersion(const InferenceEngine::Version*& versionInfo) const noexcept = 0;
+
+    /**
+     * @brief Implements deprecated API
+     */
+    // INFERENCE_ENGINE_DEPRECATED("Do not override this method. Use IE_DEFINE_EXTENSION_CREATE_FUNCTION to export extension")
```
Comment (ilyachur) [inline inference-engine/include/ie_iextension.h:193]: Comment instead of deprecation macro.

File: inference-engine/include/ie_icnn_network.hpp (hunk 10 above)
Comment (apankratovantonp) [reply inference-engine/include/ie_icnn_network.hpp:219]: Neither virtual destruction (call destructor of derived class using pointer to base class as it is intended to be used only from `std::shared_ptr`) no RTTI (use "custom RTTI" when we check result of getFunction() method to check implementation details) So virtual destructor is not needed here.

File: inference-engine/include/ie_extension.h (hunk 11 above)
Comment (apankratovantonp) [reply inference-engine/include/ie_extension.h:111]: Just to pass CI as we has some builds that use `Werror` flag. Should be uncommented before merge.

File: inference-engine/include/ie_iexecutable_network.hpp (hunk 13 above)
Comment (apankratovantonp) [reply inference-engine/include/ie_iexecutable_network.hpp:178]: no

File: inference-engine/include/ie_extension.h (hunk 11 above)
Comment (ilya-lavrenov) [reply inference-engine/include/ie_extension.h]: @apankratovantonp release branch has been created, so we are ready to merge this change. Can you enable deprecation warning back?

**Review by ilyachur — APPROVED** [review]

**Comment by azhogov** [comment]
CI passed: https://gitlab-icv.inn.intel.com/inference-engine/product-configs/merge_requests/1329
